samples/third-party-lib/src/components/ThreeDimensionalGraph/ThreeDimensionalGraphModel.ts

```diff
@@ -0,0 +1,226 @@
+    links: LinkObject[];
+}
+
+@serializable
+export default class ThreeDimensionalGraphModel extends ComponentModelBase<
+    ThreeDimensionalGraphModelProperties
+> {
+    @inject(FrameworkServiceType.BRANDING)
+    brandingService: BrandingService | undefined;
+
+    // This will be populated from the app config item reference.
+    surveyTableExtension: TableExtension | undefined;
+
+    selectedSurveyId: number | undefined;
+    hydrants: Record<string, Feature> = {};
+    surveys: Record<number, Feature> = {};
+    graphData: GraphData = { links: [], nodes: [] };
+
+    getNodeLabel = (node: NodeObject): string => {
+        if (node.type === "surveyor") {
+            return `Surveyor: ${node.id}`;
+        }
+
+        return `Hydrant Survey<br />
+        Survey ID: ${node.id}<br />
+        Hydrant number: ${node.hydrantNum}<br />
+        Result: ${node.result}<br />
+        Date: ${new Date(node.date).toLocaleString()}`;
+    };
+
+    getNodeColor = (node: NodeObject): string => {
+        if (node.type === "surveyor") {
+            return (
+                this.brandingService?.activeTheme.colors
+                    .get("primaryForeground")
+                    .toHex() ?? "white"
+            );
+        }
+
+        // Show similar color as the focus highlight on the map
+        // when the node is selected.
+        if (node.id === this.selectedSurveyId) {
+            return "teal";
+        }
+
+        if (node.result === "FAIL") {
+            return "red";
+        }
+
+        return "green";
+    };
+
+    handleNodeClick = async (node: NodeObject): Promise<void> => {
+        if (node.type === "surveyor") {
+            return;
+        }
+
+        const survey = this.surveys[node.id];
+        const hydrant = this.hydrants[node.hydrantNum];
+        if (!survey || !hydrant) {
+            return;
+        }
+
+        this.selectedSurveyId = node.id;
+        await this.messages.commands.highlights.clearFocus.execute();
+        await this.messages.commands.highlights.addFocus.execute(hydrant);
+        await this.messages.commands.results.displayDetails.execute(survey);
+    };
+
+    handleNodeHover = async (node: NodeObject | null): Promise<void> => {
+        // Nothing to do for surveyor nodes.
+        if (!node || node.type === "surveyor") {
+            return;
+        }
+
+        const hydrant = this.hydrants[node.hydrantNum];
+        if (hydrant) {
+            await this.messages.commands.highlights.pulse.execute(hydrant);
+        }
+    };
+
+    // Handle this command even when the component is inactive, which gives us
+    // an opportunity to activate it.
+    @command("three-dimensional-graph.display", { targetInactive: true })
+    protected async _handleDisplayGraph(inFeatures: Features): Promise<void> {
+        await this.messages.commands.ui.activate.execute(this);
+
+        const features = await toFeatureArray(inFeatures);
+        const hydrantFeatures = features
```

Comment:
This works, but a more robust way to verify this would be to have the hydrant layer as a property (like the survey table), and then check `feature.source === this.hydrantLayerExtension`.

Reply:
Yeah that previous logic was bugging me as it felt hacky. Features from another layer could potentially have the same field. This is much better. Thanks!